src/vsock/src/lib.rs

```diff
@@ -0,0 +1,405 @@
+            | 1 << VIRTIO_RING_F_EVENT_IDX
+            | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits()
+    }
+
+    fn protocol_features(&self) -> VhostUserProtocolFeatures {
+        VhostUserProtocolFeatures::CONFIG
+    }
+
+    fn set_event_idx(&mut self, enabled: bool) {
+        for thread in self.threads.iter() {
+            thread.lock().unwrap().event_idx = enabled;
+        }
+    }
+
+    fn update_memory(
+        &mut self,
+        atomic_mem: GuestMemoryAtomic<GuestMemoryMmap>,
+    ) -> result::Result<(), io::Error> {
+        for thread in self.threads.iter() {
+            thread.lock().unwrap().mem = Some(atomic_mem.clone());
+        }
+        Ok(())
+    }
+
+    fn handle_event(
+        &self,
+        device_event: u16,
+        evset: epoll::Events,
+        vrings: &[Arc<RwLock<Vring>>],
+        thread_id: usize,
+    ) -> result::Result<bool, io::Error> {
+        let vring_rx_lock = vrings[0].clone();
+        let vring_tx_lock = vrings[1].clone();
+
+        if evset == epoll::Events::EPOLLOUT {
+            dbg!("received epollout");
+        }
+
+        if evset != epoll::Events::EPOLLIN {
+            return Err(Error::HandleEventNotEpollIn.into());
+        }
+
+        let mut thread = self.threads[thread_id].lock().unwrap();
+        let evt_idx = thread.event_idx;
+
+        match device_event {
+            RX_QUEUE_EVENT => {
+                if thread.thread_backend.pending_rx() {
+                    thread.process_rx(vring_rx_lock, evt_idx)?;
+                }
+            }
+            TX_QUEUE_EVENT => {
+                thread.process_tx(vring_tx_lock, evt_idx)?;
+                if thread.thread_backend.pending_rx() {
+                    thread.process_rx(vring_rx_lock, evt_idx)?;
+                }
+            }
+            EVT_QUEUE_EVENT => {}
+            BACKEND_EVENT => {
+                thread.process_backend_evt(evset);
+                thread.process_tx(vring_tx_lock, evt_idx)?;
+                if thread.thread_backend.pending_rx() {
+                    thread.process_rx(vring_rx_lock, evt_idx)?;
+                }
+            }
+            _ => {
+                return Err(Error::HandleUnknownEvent.into());
+            }
+        }
+
+        Ok(false)
+    }
+
+    fn get_config(&self, _offset: u32, _size: u32) -> Vec<u8> {
+        let buf = unsafe {
+            slice::from_raw_parts(
+                &self.guest_cid as *const __u64 as *const _,
+                mem::size_of::<__u64>(),
+            )
+        };
+        buf.to_vec()
+    }
+
+    fn queues_per_thread(&self) -> Vec<u64> {
+        self.queues_per_thread.clone()
+    }
+}
+
+/// This is the public API through which an external program starts the
+/// vhost-user-vsock backend server.
+pub fn start_backend_server(vsock_config: VsockConfig) {
+    let vsock_backend = Arc::new(RwLock::new(
+        VhostUserVsockBackend::new(vsock_config.guest_cid, vsock_config.uds_path).unwrap(),
```

Comment:
what about saving entire vsock_config inside VhostUserVsockBackend ?

Reply:
Also, it would be better (consistency wise) to standardize where should keep this code. I have kept this helper in main.rs itself, either move this just like i2c or if there is a better reason to keep it here then maybe i2c should get updated to keep it here.

Also, this isn't a library really, so I won't call it lib.rs. I named this file as vhu_i2c.rs (i.e. vhost user i2c). It would be better again to standardize that ? So the entire crate looks more organized ?